File: 2022/day15.py
def dist(x0, y0, x1, y1):
    return abs(x0 - x1) + abs(y0 - y1)

# closed on both ends, NOT python-style interval
def seg(y, xs, ys, xb, yb):
    d = dist(xs, ys, xb, yb)
    dd = d - abs(ys - y)
    if dd <= 0:
        return None  # empty
    return [xs-dd, xs+dd]

import re
pat = re.compile('Sensor at x=(.+), y=(.+): closest beacon is at x=(.+), y=(.+)\n?')

def parse(ln):
    m = pat.match(ln)
    return tuple(map(int, [m.group(i) for i in range(1, 5)]))

# O(n^2); could be O(n log n) if we tried harder
def make_disjoint(segs):
    segs.sort(key=lambda s: (s[1], s[0]))
    for i in range(len(segs)):
        a = segs[i]
        if not a:
            continue
        for k in range(i+1, len(segs)):
            b = segs[k]
            if not b:
                continue
            if a[0] <= b[0] <= b[1] <= a[1]:
                b.clear()
            elif b[0] <= a[0] <= a[1] <= b[1]:
                a.clear()
                break  # to outer loop
            elif a[0] <= b[0]-1 <= a[1]:
                #a[1] = b[1]
                #b.clear()  # can't do this because breaks sort invariant
                b[0] = a[0]
                a.clear()
                break
            # other case cannot be true since we sorted above

def compute_disjoint_segs(tups, y):
    segs = [seg(y, *tup) for tup in tups]
    segs = [s for s in segs if s]  # remove Nones
    make_disjoint(segs)
    return [s for s in segs if s]  # remove empty

with open('data.txt', 'r') as file:
    y = 2000000
    #y = 10
    tups = [parse(ln) for ln in file]
    segs = compute_disjoint_segs(tups, y)
    bs = set((xb, yb) for _, _, xb, yb in tups)
    nby = sum(1 for xb, yb in bs if yb == y)
    print(-nby + sum(1+s[1]-s[0] for s in segs if s))
    # part 2
    for y in range(3000000, 4000005):
        segs = compute_disjoint_segs(tups, y)
        if len(segs) != 1:
            print(y, segs)
            # 3211051 [[-718699, 2960218], [2960220, 4308597]]
            # x = 2960219
            # >>> 4000000 * 2960219 + 3211051
# 11840879211051

File: 2022-go/day15.go
package main

import (
	"bufio"
	"cmp"
	"fmt"
	"math"
	"os"
	"regexp"
	"slices"
	"strconv"
)

var sensorRe = regexp.MustCompile(
	`Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)`)

func parseLine(line string) [4]int {
	var out [4]int
	m := sensorRe.FindStringSubmatch(line)
	for i := 0; i < 4; i++ {
		out[i], _ = strconv.Atoi(m[i+1])
	}
	return out
}

func parseFile(filename string) *[][4]int {
	fin, _ := os.Open(filename)
	defer fin.Close()
	sc := bufio.NewScanner(bufio.NewReader(fin))
	var out [][4]int
	for sc.Scan() {
		out = append(out, parseLine(sc.Text()))
	}
	return &out
}

func dist(x0 int, y0 int, x1 int, y1 int) int {
	return abs(x0-x1) + abs(y0-y1)
}

func seg(y int, xs int, ys int, xb int, yb int) ([2]int, bool) {
	d := dist(xs, ys, xb, yb)
	dd := d - abs(ys-y)
	if dd <= 0 {
		return [2]int{0, 0}, false
	}
	return [2]int{xs - dd, xs + dd}, true
}

// apparently no need (*seg) here because [2] not []
func setClear(seg *[2]int) {
	seg[0] = math.MinInt
	seg[1] = math.MinInt
}

func isClear(seg [2]int) bool {
	return seg[0] == math.MinInt && seg[1] == math.MinInt
}

func comPairBackwards(a [2]int, b [2]int) int {
	c := cmp.Compare(a[1], b[1])
	if c != 0 {
		return c
	} else {
		return cmp.Compare(a[0], b[0])
	}
}

func lessEq(a int, b int, c int, d int) bool {
	return a <= b && b <= c && c <= d
}

// O(n^2); could be O(n log n) if we tried harder using more sophisticated
// interval algos
func makeDisjoint(segs *[][2]int) {
	slices.SortFunc(*segs, comPairBackwards)
	for i := range *segs {
		a := &(*segs)[i]
		if isClear(*a) {
			continue
		}
		for k := i + 1; k < len(*segs); k++ {
			b := &(*segs)[k]
			if isClear(*b) {
				continue
			}
			if lessEq(a[0], b[0], b[1], a[1]) {
				setClear(b)
			} else if lessEq(b[0], a[0], a[1], b[1]) {
				setClear(a)
				break
			} else if a[0]+1 <= b[0] && b[0] <= a[1]+1 {
				// a[1] = b[1]
				// b.clear()  // cannot do this because breaks sort invariant
				b[0] = a[0]
				setClear(a)
				break
			}
			// other case cannot be true since we sorted above
		}
	}
}

func computeDisjointSegs(tups *[][4]int, y int) *[][2]int {
	var segs [][2]int
	for _, tup := range *tups {
		s, ok := seg(y, tup[0], tup[1], tup[2], tup[3])
		if ok {
			segs = append(segs, s)
		}
	}
	makeDisjoint(&segs)
	out := slices.DeleteFunc(segs, isClear)
	return &out
}

func main() {
	filename := "/home/xdavidliu/Documents/aoc/input15.txt"
	tups := parseFile(filename)
	y := 2_000_000
	segs := computeDisjointSegs(tups, y)
	bys := make(set[int])
	for _, tup := range *tups {
		xb, yb := tup[2], tup[3]
		if y == yb {
			bys.insert(xb)
		}
	}
	nby := len(bys)
	total := 0
	for _, s := range *segs {
		total += 1 + s[1] - s[0]
	}
	fmt.Println("part 1 =", total-nby)        // 5394423
	fmt.Println("part 2 =", solvePart2(tups)) // 11840879211051
}

func solvePart2(tups *[][4]int) int {
	for y := 3_000_000; y < 4_000_005; y++ {
		segs := computeDisjointSegs(tups, y)
		if len(*segs) != 1 {
			xb := (*segs)[1][0] - 1
			return 4_000_000*xb + y
		}
	}
	panic("part 2")
}
